allennlp/nn/beam_search.py

```diff
@@ -381,3 +403,152 @@ def _search(
+        start_predictions : `torch.Tensor`
+            A tensor containing the initial predictions with shape `(batch_size,)`.
+            Usually the initial predictions are just the index of the "start" token
+            in the target vocabulary.
+        start_state : `StateType`
+            The initial state passed to the `step` function. Each value of the state dict
+            should be a tensor of shape `(batch_size, *)`, where `*` means any other
+            number of dimensions.
+        step : `StepFunctionType`
+            A function that is responsible for computing the next most likely tokens,
+            given the current state and the predictions from the last time step.
+            The function should accept two arguments. The first being a tensor
+            of shape `(group_size,)`, representing the index of the predicted
+            tokens from the last time step, and the second being the current state.
+            The `group_size` will be `batch_size * beam_size`, except in the initial
+            step, for which it will just be `batch_size`.
+            The function is expected to return a tuple, where the first element
+            is a tensor of shape `(group_size, target_vocab_size)` containing
+            the log probabilities of the tokens for the next step, and the second
+            element is the updated state. The tensor in the state should have shape
+            `(group_size, *)`, where `*` means any other number of dimensions.
+        max_steps : `int`, optional (default = `50`)
+            The maximum number of decoding steps to take, i.e. the maximum length
+            of the predicted sequences.
+        beam_size : `int`, optional (default = `10`)
+            The width of the beam used.
+        p : `float`, optional (default = `0.9`)
+            The cumulative probability of top next tokens to select from.
+        temperature : `float`, optional (default = `1.0`)
+            The 'temperature' of the probability distribution to be sampled
+            from, lowering below 1.0 creates a sharper distribution.
+
+        # Returns
+
+        `Tuple[torch.Tensor, torch.Tensor]`
+            Tuple of `(predictions, log_probabilities)`, where `predictions`
+            has shape `(batch_size, beam_size, max_steps)` and `log_probabilities`
+            has shape `(batch_size, beam_size)`.
+        """
+        # Make sure `p` is a valid cumulative probability threshold.
+        if type(p) is not float or p < 0.0 or p > 1.0:
+            raise ConfigurationError(
+                f'{"value of cumulative probability threshold `p`=({p}) too small."}'
+                f'{"`p` must be a float between `0.0` and `1.0`"}'
+            )
+
+        sampler_p = TopPSampler(p, temperature)
+        return cls(
+            end_index=end_index,
+            max_steps=max_steps,
+            beam_size=beam_size,
+            per_node_beam_size=1,
+            sampler=sampler_p,
+        )
+
+
```

Comment:
We might also need a default registered constructor. For example, with the `Vocabulary` object, the default is `from_instances`: https://github.com/allenai/allennlp/blob/ae7cf85b8f755f086341a087aeca1ba7c1df3769/allennlp/data/vocabulary.py#L223
which has to be registered with the others here: https://github.com/allenai/allennlp/blob/ae7cf85b8f755f086341a087aeca1ba7c1df3769/allennlp/data/vocabulary.py#L766.